race condition in system tests -- table list, then get

https://source.cloud.google.com/results/invocations/ed727aae-0f39-444a-ae80-525b801f19cd/targets/cloud-devrel%2Fclient-libraries%2Fpython%2Fgoogleapis%2Fpython-bigquery-sqlalchemy%2Fpresubmit%2Fpresubmit/log

```
______________________________ test_tables_list _______________________________

engine = Engine(bigquery://)
engine_using_test_dataset = Engine(bigquery:///test_pybigquery)

    def test_tables_list(engine, engine_using_test_dataset):
>       tables = engine.table_names()

tests/system/test_sqlalchemy_bigquery.py:355:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.nox/system-3-8/lib/python3.8/site-packages/sqlalchemy/engine/base.py:2317: in table_names
    return self.dialect.get_table_names(conn, schema)
pybigquery/sqlalchemy_bigquery.py:592: in get_table_names
    return self._get_table_or_view_names(connection, "TABLE", schema)
pybigquery/sqlalchemy_bigquery.py:395: in _get_table_or_view_names
    for table in tables:
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/page_iterator.py:212: in _items_iter
    for page in self._page_iter(increment=False):
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/page_iterator.py:243: in _page_iter
    page = self._next_page()
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/page_iterator.py:369: in _next_page
    response = self._get_next_page_response()
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/page_iterator.py:418: in _get_next_page_response
    return self.api_request(
.nox/system-3-8/lib/python3.8/site-packages/google/cloud/bigquery/client.py:1330: in api_request
    return self._call_api(
.nox/system-3-8/lib/python3.8/site-packages/google/cloud/bigquery/client.py:640: in _call_api
    return call()
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/retry.py:281: in retry_wrapped_func
    return retry_target(
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/retry.py:184: in retry_target
    return target()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <google.cloud.bigquery._http.Connection object at 0x7fc6a7d07310>
method = 'GET'
path = '/projects/precise-truck-742/datasets/create_table_1617200732845/tables'
query_params = {}, data = None, content_type = None, headers = None
api_base_url = None, api_version = None, expect_json = True
_target_object = None, timeout = None

    def api_request(
        self,
        method,
        path,
        query_params=None,
        data=None,
        content_type=None,
        headers=None,
        api_base_url=None,
        api_version=None,
        expect_json=True,
        _target_object=None,
        timeout=_DEFAULT_TIMEOUT,
    ):
        """Make a request over the HTTP transport to the API.

        You shouldn't need to use this method, but if you plan to
        interact with the API using these primitives, this is the
        correct one to use.

        :type method: str
        :param method: The HTTP method name (ie, ``GET``, ``POST``, etc).
                       Required.

        :type path: str
        :param path: The path to the resource (ie, ``'/b/bucket-name'``).
                     Required.

        :type query_params: dict or list
        :param query_params: A dictionary of keys and values (or list of
                             key-value pairs) to insert into the query
                             string of the URL.

        :type data: str
        :param data: The data to send as the body of the request. Default is
                     the empty string.

        :type content_type: str
        :param content_type: The proper MIME type of the data provided. Default
                             is None.

        :type headers: dict
        :param headers: extra HTTP headers to be sent with the request.

        :type api_base_url: str
        :param api_base_url: The base URL for the API endpoint.
                             Typically you won't have to provide this.
                             Default is the standard API base URL.

        :type api_version: str
        :param api_version: The version of the API to call.  Typically
                            you shouldn't provide this and instead use
                            the default for the library.  Default is the
                            latest API version supported by
                            google-cloud-python.

        :type expect_json: bool
        :param expect_json: If True, this method will try to parse the
                            response as JSON and raise an exception if
                            that cannot be done.  Default is True.

        :type _target_object: :class:`object`
        :param _target_object:
            (Optional) Protected argument to be used by library callers. This
            can allow custom behavior, for example, to defer an HTTP request
            and complete initialization of the object at a later time.

        :type timeout: float or tuple
        :param timeout: (optional) The amount of time, in seconds, to wait
            for the server response.

            Can also be passed as a tuple (connect_timeout, read_timeout).
            See :meth:`requests.Session.request` documentation for details.

        :raises ~google.cloud.exceptions.GoogleCloudError: if the response code
            is not 200 OK.
        :raises ValueError: if the response content type is not JSON.
        :rtype: dict or str
        :returns: The API response payload, either as a raw string or
                  a dictionary if the response is valid JSON.
        """
        url = self.build_api_url(
            path=path,
            query_params=query_params,
            api_base_url=api_base_url,
            api_version=api_version,
        )

        # Making the executive decision that any dictionary
        # data will be sent properly as JSON.
        if data and isinstance(data, dict):
            data = json.dumps(data)
            content_type = "application/json"

        response = self._make_request(
            method=method,
            url=url,
            data=data,
            content_type=content_type,
            headers=headers,
            target_object=_target_object,
            timeout=timeout,
        )

        if not 200 <= response.status_code < 300:
>           raise exceptions.from_http_response(response)
E           google.api_core.exceptions.NotFound: 404 GET https://bigquery.googleapis.com/bigquery/v2/projects/precise-truck-742/datasets/create_table_1617200732845/tables?prettyPrint=false: Not found: Dataset precise-truck-742:create_table_1617200732845
```

```
_______________________________ test_view_names ________________________________

inspector = <sqlalchemy.engine.reflection.Inspector object at 0x7fc6a7638850>
inspector_using_test_dataset = <sqlalchemy.engine.reflection.Inspector object at 0x7fc6a76382b0>

    def test_view_names(inspector, inspector_using_test_dataset):
>       view_names = inspector.get_view_names()

tests/system/test_sqlalchemy_bigquery.py:592:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.nox/system-3-8/lib/python3.8/site-packages/sqlalchemy/engine/reflection.py:325: in get_view_names
    return self.dialect.get_view_names(
pybigquery/sqlalchemy_bigquery.py:598: in get_view_names
    return self._get_table_or_view_names(connection, "VIEW", schema)
pybigquery/sqlalchemy_bigquery.py:395: in _get_table_or_view_names
    for table in tables:
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/page_iterator.py:212: in _items_iter
    for page in self._page_iter(increment=False):
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/page_iterator.py:243: in _page_iter
    page = self._next_page()
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/page_iterator.py:369: in _next_page
    response = self._get_next_page_response()
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/page_iterator.py:418: in _get_next_page_response
    return self.api_request(
.nox/system-3-8/lib/python3.8/site-packages/google/cloud/bigquery/client.py:1330: in api_request
    return self._call_api(
.nox/system-3-8/lib/python3.8/site-packages/google/cloud/bigquery/client.py:640: in _call_api
    return call()
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/retry.py:281: in retry_wrapped_func
    return retry_target(
.nox/system-3-8/lib/python3.8/site-packages/google/api_core/retry.py:184: in retry_target
    return target()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <google.cloud.bigquery._http.Connection object at 0x7fc6a7d07310>
method = 'GET'
path = '/projects/precise-truck-742/datasets/bq_system_test_1617200758917/tables'
query_params = {}, data = None, content_type = None, headers = None
api_base_url = None, api_version = None, expect_json = True
_target_object = None, timeout = None

    def api_request(
        self,
        method,
        path,
        query_params=None,
        data=None,
        content_type=None,
        headers=None,
        api_base_url=None,
        api_version=None,
        expect_json=True,
        _target_object=None,
        timeout=_DEFAULT_TIMEOUT,
    ):
        """Make a request over the HTTP transport to the API.

        You shouldn't need to use this method, but if you plan to
        interact with the API using these primitives, this is the
        correct one to use.

        :type method: str
        :param method: The HTTP method name (ie, ``GET``, ``POST``, etc).
                       Required.

        :type path: str
        :param path: The path to the resource (ie, ``'/b/bucket-name'``).
                     Required.

        :type query_params: dict or list
        :param query_params: A dictionary of keys and values (or list of
                             key-value pairs) to insert into the query
                             string of the URL.

        :type data: str
        :param data: The data to send as the body of the request. Default is
                     the empty string.

        :type content_type: str
        :param content_type: The proper MIME type of the data provided. Default
                             is None.

        :type headers: dict
        :param headers: extra HTTP headers to be sent with the request.

        :type api_base_url: str
        :param api_base_url: The base URL for the API endpoint.
                             Typically you won't have to provide this.
                             Default is the standard API base URL.

        :type api_version: str
        :param api_version: The version of the API to call.  Typically
                            you shouldn't provide this and instead use
                            the default for the library.  Default is the
                            latest API version supported by
                            google-cloud-python.

        :type expect_json: bool
        :param expect_json: If True, this method will try to parse the
                            response as JSON and raise an exception if
                            that cannot be done.  Default is True.

        :type _target_object: :class:`object`
        :param _target_object:
            (Optional) Protected argument to be used by library callers. This
            can allow custom behavior, for example, to defer an HTTP request
            and complete initialization of the object at a later time.

        :type timeout: float or tuple
        :param timeout: (optional) The amount of time, in seconds, to wait
            for the server response.

            Can also be passed as a tuple (connect_timeout, read_timeout).
            See :meth:`requests.Session.request` documentation for details.

        :raises ~google.cloud.exceptions.GoogleCloudError: if the response code
            is not 200 OK.
        :raises ValueError: if the response content type is not JSON.
        :rtype: dict or str
        :returns: The API response payload, either as a raw string or
                  a dictionary if the response is valid JSON.
        """
        url = self.build_api_url(
            path=path,
            query_params=query_params,
            api_base_url=api_base_url,
            api_version=api_version,
        )

        # Making the executive decision that any dictionary
        # data will be sent properly as JSON.
        if data and isinstance(data, dict):
            data = json.dumps(data)
            content_type = "application/json"

        response = self._make_request(
            method=method,
            url=url,
            data=data,
            content_type=content_type,
            headers=headers,
            target_object=_target_object,
            timeout=timeout,
        )

        if not 200 <= response.status_code < 300:
>           raise exceptions.from_http_response(response)
E           google.api_core.exceptions.NotFound: 404 GET https://bigquery.googleapis.com/bigquery/v2/projects/precise-truck-742/datasets/bq_system_test_1617200758917/tables?prettyPrint=false: Not found: Dataset precise-truck-742:bq_system_test_1617200758917

.nox/system-3-8/lib/python3.8/site-packages/google/cloud/_http.py:483: NotFound
```

These tables / datasets belong to other libraries. I think there must a race condition between list and getting additional table details. Also, if all that's required is the table name, the get may be unnecessary.

--- a/pybigquery/sqlalchemy_bigquery.py
+++ b/pybigquery/sqlalchemy_bigquery.py
@@ -26,7 +26,6 @@ import operator
 
 from google import auth
 import google.api_core.exceptions
-from google.cloud import bigquery
 from google.cloud.bigquery import dbapi
 from google.cloud.bigquery.schema import SchemaField
 from google.cloud.bigquery.table import TableReference

--- a/pybigquery/sqlalchemy_bigquery.py
+++ b/pybigquery/sqlalchemy_bigquery.py
@@ -25,6 +25,7 @@ from __future__ import unicode_literals
 import operator
 
 from google import auth
+import google.api_core.exceptions
 from google.cloud import bigquery
 from google.cloud.bigquery import dbapi
 from google.cloud.bigquery.schema import SchemaField
@@ -434,10 +435,17 @@ class BigQueryDialect(DefaultDialect):
             if current_schema is not None and current_schema != dataset.dataset_id:
                 continue
 
-            tables = client.list_tables(dataset.reference)
-            for table in tables:
-                if table_type == table.table_type:
-                    result.append(get_table_name(table))
+            try:
+                tables = client.list_tables(dataset.reference)
+                for table in tables:
+                    if table_type == table.table_type:
+                        result.append(get_table_name(table))
+            except google.api_core.exceptions.NotFound:
+                # It's possible that the dataset was deleted between when we
+                # fetched the list of datasets and when we try to list the
+                # tables from it. See:
+                # https://github.com/googleapis/python-bigquery-sqlalchemy/issues/105
+                pass
         return result
 
     @staticmethod

--- a/tests/unit/test_sqlalchemy_bigquery.py
+++ b/tests/unit/test_sqlalchemy_bigquery.py
@@ -1,0 +1,13 @@
+# Copyright 2021 The PyBigQuery Authors
+#
+# Use of this source code is governed by an MIT-style
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+
+
+def test_get_table_names():
+    assert False
+
+
+def test_get_view_names():
+    assert False

--- a/tests/unit/test_sqlalchemy_bigquery.py
+++ b/tests/unit/test_sqlalchemy_bigquery.py
@@ -4,10 +4,136 @@
 # license that can be found in the LICENSE file or at
 # https://opensource.org/licenses/MIT.
 
+from unittest import mock
 
-def test_get_table_names():
-    assert False
+import google.api_core.exceptions
+from google.cloud import bigquery
+from google.cloud.bigquery.dataset import DatasetListItem
+from google.cloud.bigquery.table import TableListItem
+import pytest
+import sqlalchemy
 
 
-def test_get_view_names():
-    assert False
+@pytest.fixture
+def mock_bigquery_client():
+    return mock.create_autospec(bigquery.Client, instance=True)
+
+
+@pytest.fixture
+def mock_connection(monkeypatch, mock_bigquery_client):
+    from pybigquery import sqlalchemy_bigquery
+
+    def mock_connect_args(*args, **kwargs):
+        return ([mock_bigquery_client], {})
+
+    monkeypatch.setattr(
+        sqlalchemy_bigquery.BigQueryDialect, "create_connect_args", mock_connect_args
+    )
+
+
+@pytest.fixture
+def engine_under_test(mock_connection):
+    return sqlalchemy.create_engine("bigquery://")
+
+
+@pytest.fixture
+def inspector_under_test(engine_under_test):
+    from sqlalchemy.engine.reflection import Inspector
+
+    return Inspector.from_engine(engine_under_test)
+
+
+def dataset_item(dataset_id):
+    return DatasetListItem(
+        {"datasetReference": {"projectId": "some-project-id", "datasetId": dataset_id}}
+    )
+
+
+def table_item(dataset_id, table_id, type_="TABLE"):
+    return TableListItem(
+        {
+            "type": type_,
+            "tableReference": {
+                "projectId": "some-project-id",
+                "datasetId": dataset_id,
+                "tableId": table_id,
+            },
+        }
+    )
+
+
+@pytest.mark.parametrize(
+    ["datasets_list", "tables_lists", "expected"],
+    [
+        ([], [], []),
+        ([dataset_item("dataset_1")], [[]], []),
+        (
+            [dataset_item("dataset_1"), dataset_item("dataset_2")],
+            [
+                [table_item("dataset_1", "d1t1"), table_item("dataset_1", "d1t2")],
+                [
+                    table_item("dataset_2", "d2t1"),
+                    table_item("dataset_2", "d2view", type_="VIEW"),
+                ],
+            ],
+            ["dataset_1.d1t1", "dataset_1.d1t2", "dataset_2.d2t1"],
+        ),
+        (
+            [dataset_item("dataset_1"), dataset_item("dataset_deleted")],
+            [
+                [table_item("dataset_1", "d1t1")],
+                google.api_core.exceptions.NotFound("dataset_deleted"),
+            ],
+            ["dataset_1.d1t1"],
+        ),
+    ],
+)
+def test_get_table_names(
+    engine_under_test, mock_bigquery_client, datasets_list, tables_lists, expected
+):
+    mock_bigquery_client.list_datasets.return_value = datasets_list
+    mock_bigquery_client.list_tables.side_effect = tables_lists
+    table_names = engine_under_test.table_names()
+    mock_bigquery_client.list_datasets.assert_called_once()
+    assert mock_bigquery_client.list_tables.call_count == len(datasets_list)
+    assert list(sorted(table_names)) == list(sorted(expected))
+
+
+@pytest.mark.parametrize(
+    ["datasets_list", "tables_lists", "expected"],
+    [
+        ([], [], []),
+        ([dataset_item("dataset_1")], [[]], []),
+        (
+            [dataset_item("dataset_1"), dataset_item("dataset_2")],
+            [
+                [
+                    table_item("dataset_1", "d1t1"),
+                    table_item("dataset_1", "d1view", type_="VIEW"),
+                ],
+                [
+                    table_item("dataset_2", "d2t1"),
+                    table_item("dataset_2", "d2view", type_="VIEW"),
+                ],
+            ],
+            ["dataset_1.d1view", "dataset_2.d2view"],
+        ),
+        (
+            [dataset_item("dataset_1"), dataset_item("dataset_deleted")],
+            [
+                [table_item("dataset_1", "d1view", type_="VIEW")],
+                google.api_core.exceptions.NotFound("dataset_deleted"),
+            ],
+            ["dataset_1.d1view"],
+        ),
+    ],
+)
+def test_get_view_names(
+    inspector_under_test, mock_bigquery_client, datasets_list, tables_lists, expected
+):
+    mock_bigquery_client.list_datasets.return_value = datasets_list
+    mock_bigquery_client.list_tables.side_effect = tables_lists
+    view_names = inspector_under_test.get_view_names()
+    mock_bigquery_client.list_datasets.assert_called_once()
+    assert mock_bigquery_client.list_tables.call_count == len(datasets_list)
+    assert list(sorted(view_names)) == list(sorted(expected))

